src/docfx.website.themes/common/common.js

```diff
@@ -24,83 +24,126 @@ function getDirectoryName(path) {
+            url += '/new';
+            url += '/' + branch;
+            url += '/' + getOverrideFolder(path);
+            url += '/new?filename=' + getHtmlId(uid) + '.md';
+            url += '&value=' + encodeURIComponent(getOverrideTemplate(uid));
+            return url;
         }
-        var path = getGithubUrlPrefix(repo);
-        if (path != '') {
-            path += '/blob' + '/' + branch + '/' + remote.path;
-            if (linenum > 0) path += '/#L' + linenum;
+    },
+    'vso': {
+        'testRegex': /^https:\/\/.*\.visualstudio\.com\/.*/i,
+        'generateUrl': function(repo, branch, path, startLine) {
+            var url =  repo + '?path=' + path + '&version=GB' + branch;
+            if (startLine && startLine > 0) {
+                url += '&line=' + startLine;
+            }
+            return url;
+        },
+        'generateNewFileUrl': function(repo, branch, path, uid) {
+            return '';
         }
-        return path;
-    } else {
-        return '';
     }
 }
 
-function getGithubUrlPrefix(repo) {
-    var regex = /^(?:https?:\/\/)?(?:\S+\@)?(?:\S+\.)?(github\.com(?:\/|:).*)/gi;
-    if (!regex.test(repo)) return '';
-    return repo.replace(regex, function (match, p1, offset, string) {
-        return 'https://' + p1.replace(':', '/');
-    })
+function normalizeGitUrlToHttps(repo) {
+    var pos = repo.indexOf('@');
+    if (pos == -1) return repo;
+    return 'https://' + repo.substr(pos + 1).replace(/:/g, '/');
+}
+
+function getNewFileUrl(uid, gitContribute, gitUrlPattern) {
+    // do not support VSO for now
+    if (!gitContribute || !gitContribute.repo) return '';
+    var repo = gitContribute.repo;
+    if (repo.substr(-4) === '.git') {
+        repo = repo.substr(0, repo.length - 4);
+    }
+
+    var patternName = getPatternName(repo, gitUrlPattern);
+    if (!patternName) return patternName;
+    return gitUrlPatternItems[patternName].generateNewFileUrl(repo, gitContribute.branch, gitContribute.path, uid);
+}
+
+function getRemoteUrl(remote, startLine, gitContribute, gitUrlPattern) {
+    if (!remote || !remote.repo) return '';
+    var repo = remote.repo;
+    if (gitContribute && gitContribute.repo) repo = gitContribute.repo;
+    var branch = remote.branch;
+    if (gitContribute && gitContribute.branch) branch = gitContribute.branch;
+    if (repo.substr(-4) === '.git') {
+        repo = repo.substr(0, repo.length - 4);
+    }
+
+    var patternName = getPatternName(repo, gitUrlPattern);
+    if (!patternName) return '';
+    return gitUrlPatternItems[patternName].generateUrl(repo, branch, remote.path, startLine);
+}
+
+function getPatternName(repo, gitUrlPattern)
+{
+    var patternName = '';
+    if (gitUrlPattern && gitUrlPattern.toLowerCase() in gitUrlPatternItems) {
+        patternName = gitUrlPattern.toLowerCase();
+    } else {
+        for (var p in gitUrlPatternItems) {
+            if (gitUrlPatternItems[p].testRegex.test(repo)) {
+                patternName = p;
```

Comment:
sounds nice

---
In reply to: [78559120](https://github.com/dotnet/docfx/pull/696#discussion_r78559120) [](ancestors = 78559120)